### PR TITLE
quic: Add QUIC App Layer - v7

### PR DESCRIPTION
**Pull request by sharksforarms**
Previous PR: https://github.com/OISF/suricata/pull/6313

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/3440

Describe changes:
- Parses gquic version Q043-Q046 and logs a CYU hash
- Added a mix of unit and suricata-verify pcap tests
- Fails gracefully with unknown QUIC versions
- Other versions can be added later w/ pcaps + testing
- Logs a cyu hash and string to eve.json
  - Example ```{"timestamp":"2020-09-02T15:58:21.139196+0000","flow_id":1610311033233340,"pcap_cnt":1,"event_type":"quic","src_ip":"10.0.2.15","src_port":51989,"dest_ip":"172.217.164.142","dest_port":443,"proto":"UDP","quic":[{"cyu":{"hash":"910a5e3a4d51593bd59a44611544f209","string":"46,PAD-SNI-VER-CCS-UAID-TCID-PDMD-SMHL-ICSL-NONP-MIDS-SCLS-CSCT-COPT-IRTT-CFCW-SFCW"}]}}```

References:
   - https://docs.google.com/document/d/1WJvyZflAO2pq77yOLbp9NsGjC1CHetAXV8I0fQe-B_U/edit
   - https://docs.google.com/document/d/1g5nIXAIkN_Y-7XJW5K45IblHd_L2f5LTaDUDwvZ5L6g/edit
   - https://www.slideshare.net/shigeki_ohtsu/quic-overview
   - https://tools.ietf.org/html/draft-ietf-quic-transport-17#section-19.8
   - https://github.com/salesforce/GQUIC_Protocol_Analyzer/blob/master/src/gquic-protocol.pac
   - cyu: https://engineering.salesforce.com/gquic-protocol-analysis-and-fingerprinting-in-zeek-a4178855d75f

suricata-verify-pr: 525

### CODE REVIEW

File: src/detect-quic-version.c [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    sigmatch_table[DETECT_AL_QUIC_VERSION].Setup = DetectQuicVersionSetup;
+    sigmatch_table[DETECT_AL_QUIC_VERSION].Free = DetectQuicVersionFree;
+#ifdef UNITTESTS
+    sigmatch_table[DETECT_AL_QUIC_VERSION].RegisterTests = DetectQuicVersionRegisterTests;
+#endif
+
+    DetectAppLayerInspectEngineRegister2("quic.version", ALPROTO_QUIC, SIG_FLAG_TOSERVER, 1,
+            DetectEngineInspectQuicVersionGeneric, NULL);
+
+    quic_version_id = DetectBufferTypeGetByName("quic.version");
```
Comment (catenacyber) [inline src/detect-quic-version.c:68]: @victorjulien do we need this ?
This is not a sticky buffer, so why would we need `DetectBufferTypeGetByName ` ?

File: src/detect-quic-cyu-string.c [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+                DETECT_CI_FLAGS_SINGLE, DETECT_ENGINE_CONTENT_INSPECTION_MODE_STATE);
+        if (match == 1) {
+            return DETECT_ENGINE_INSPECT_SIG_MATCH;
+        }
+        local_id++;
+    }
+    return DETECT_ENGINE_INSPECT_SIG_NO_MATCH;
+}
+
+typedef struct PrefilterMpmQuicString {
```
Comment (catenacyber) [inline src/detect-quic-cyu-string.c:119]: You can reuse `PrefilterMpmListId`

File: src/detect-quic-cyu-hash.c [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+    sigmatch_table[DETECT_AL_QUIC_CYU_HASH].desc = "sticky buffer to match on the QUIC CYU hash";
+    sigmatch_table[DETECT_AL_QUIC_CYU_HASH].url = "/rules/" KEYWORD_DOC;
+    sigmatch_table[DETECT_AL_QUIC_CYU_HASH].Setup = DetectQuicCyuHashSetup;
+    sigmatch_table[DETECT_AL_QUIC_CYU_HASH].flags |= SIGMATCH_NOOPT;
+#ifdef UNITTESTS
+    sigmatch_table[DETECT_AL_QUIC_CYU_HASH].RegisterTests = DetectQuicCyuHashRegisterTests;
+#endif
+
+    DetectAppLayerMpmRegister2(
+            BUFFER_NAME, SIG_FLAG_TOSERVER, 2, PrefilterMpmQuicHashRegister, NULL, ALPROTO_QUIC, 1);
```
Comment (catenacyber) [inline src/detect-quic-cyu-hash.c:230]: I think you can use generic `PrefilterGenericMpmRegister` cf https://github.com/OISF/suricata/pull/6270#discussion_r700104353

File: rust/src/quic/quic.rs [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+
+impl Drop for QuicTransaction {
+    fn drop(&mut self) {
+        self.free();
+    }
+}
+
+pub struct QuicState {
+    tx_id: u64,
+    pub version: u32,
```
Comment (catenacyber) [inline rust/src/quic/quic.rs:72]: Should this `version` rather be a field of the transaction structure ?

File: rust/src/quic/quic.rs [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+}
+
+impl Drop for QuicTransaction {
+    fn drop(&mut self) {
+        self.free();
+    }
+}
+
+pub struct QuicState {
+    tx_id: u64,
```
Comment (catenacyber) [inline rust/src/quic/quic.rs:71]: Is this not simply `transactions.len()` ?

Comment (catenacyber) [reply rust/src/quic/quic.rs:71]: Oh no, never mind my comment

**Review by catenacyber — APPROVED** [review]
Looks good to me, even if there are some nits about style

File: rust/src/quic/quic.rs (hunk 5 above)
Comment (victorjulien) [reply rust/src/quic/quic.rs:71]: we should probably rename all these fields and the API too, to indicate we're tracking the max_id

File: src/detect-quic-cyu-hash.c (hunk 3 above)
Comment (victorjulien) [reply src/detect-quic-cyu-hash.c:230]: we don't have a generic one for multi-buffers yet

File: src/detect-quic-version.c [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+ * \retval 1 match.
+ */
+static int DetectQuicVersionMatch(DetectEngineThreadCtx *det_ctx, Flow *f, uint8_t flags,
+        void *state, void *txv, const Signature *s, const SigMatchCtx *ctx)
+{
+    const DetectU32Data *de = (const DetectU32Data *)ctx;
+    uint32_t version;
+
+    version = rs_quic_tx_get_version(state);
+    SCLogNotice("QUIC: %d", version);
```
Comment (victorjulien) [inline src/detect-quic-version.c:101]: debug leftover

File: src/detect-quic-version.c (hunk 1 above)
Comment (victorjulien) [reply src/detect-quic-version.c:68]: "Buffer" is a bit of a misnomer. Its registering a inspection list id. You can see that the id is used in `DetectQuicVersionSetup` below

File: src/output-json-quic.c [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    return TM_ECODE_FAILED;
+}
+
+static TmEcode JsonQuicLogThreadDeinit(ThreadVars *t, void *data)
+{
+    JsonQuicLogThread *thread = (JsonQuicLogThread *)data;
+    if (thread == NULL) {
+        return TM_ECODE_OK;
+    }
+    SCFree(thread);
```
Comment (victorjulien) [inline src/output-json-quic.c:139]: missing free of `thread->ctx` -> `FreeEveThreadCtx(thread->ctx);`

**Comment by victorjulien** [comment]
Think this needs a few more small adjustments, but overall looks good and almost ready for merge.

The biggest thing missing is documentation of the detection keywords.

File: src/detect-quic-version.c (hunk 1 above)
Comment (catenacyber) [reply src/detect-quic-version.c:68]: Thanks Victor, what is an inspection list id ?

**Comment by sharksforarms** [comment]
Thanks for the review, I should have a bit of time this week to update it. I'm also interested in getting the fuzzing running, any advice for this?

**Comment by catenacyber** [comment]
> Thanks for the review, I should have a bit of time this week to update it. I'm also interested in getting the fuzzing running, any advice for this?

There is a README in the fuzz directory https://github.com/OISF/suricata/blob/master/src/tests/fuzz/README
I recommend you run target `fuzz_applayerparserparse` on QUIC only, so by renaming it to `fuzz_applayerparserparse_quic` or using env variable `FUZZ_APPLAYER=30`
As it is rust, I recommend that you set `RUSTFLAGS` with what you need cf https://github.com/google/oss-fuzz/blob/master/projects/suricata/build.sh#L75

**Comment by sharksforarms** [comment]
Continued in: https://github.com/OISF/suricata/pull/6406

**Comment by sharksforarms** [comment]
> > Thanks for the review, I should have a bit of time this week to update it. I'm also interested in getting the fuzzing running, any advice for this?
> 
> There is a README in the fuzz directory https://github.com/OISF/suricata/blob/master/src/tests/fuzz/README
> I recommend you run target `fuzz_applayerparserparse` on QUIC only, so by renaming it to `fuzz_applayerparserparse_quic` or using env variable `FUZZ_APPLAYER=30`
> As it is rust, I recommend that you set `RUSTFLAGS` with what you need cf https://github.com/google/oss-fuzz/blob/master/projects/suricata/build.sh#L75

Are these instructions up to date? I tried the oss-fuzz and got `FUZZING_LANGUAGE: unbound variable` also tried the libfuzzer, but the target just terminates without output...

**Comment by catenacyber** [comment]
> Are these instructions up to date?also tried the libfuzzer, but the target just terminates without output...

They should be...
Cf https://github.com/OISF/suricata/pull/6407
Thanks

>  I tried the oss-fuzz and got FUZZING_LANGUAGE: unbound variable 

I think you now need to get the latest oss-fuzz like `git clone  --depth 1 https://github.com/google/oss-fuzz` instead of the old specific branch

> also tried the libfuzzer, but the target just terminates without output...

What is your `configure`command output ?
You can try the more complete flags proposition from #6407

**Pull request closed without merging**